### PR TITLE
Write lists of minions targeted by syndic masters to job cache

### LINKED ISSUE #30528: Missing Minion notifications missing from job cache
Working with Salt with a syndic the notification that a minion did not return was missing from the job cache in the master of masters, while the same job run from the syndic directly would show up in the job cache on the syndic.

Doing some testing on 2015.8.3 (the versions report is below) I didn't see the missing minion info in either of the job caches. It would be good to make sure that info shows up in the job cache and not just on the cli.

```
local:
    Salt Version:
               Salt: 2015.8.3
     
    Dependency Versions:
             Jinja2: 2.7.2
           M2Crypto: Not Installed
               Mako: 0.9.1
             PyYAML: 3.10
              PyZMQ: 14.0.1
             Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
               RAET: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.0.4
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: 1.5
              gitdb: 0.5.4
          gitpython: 0.3.2 RC1
              ioflo: Not Installed
            libnacl: Not Installed
       msgpack-pure: Not Installed
     msgpack-python: 0.3.0
       mysql-python: 1.2.3
          pycparser: Not Installed
           pycrypto: 2.6.1
             pygit2: Not Installed
       python-gnupg: Not Installed
              smmap: 0.8.2
            timelib: Not Installed
     
    System Versions:
               dist: Ubuntu 14.04 trusty
            machine: x86_64
            release: 3.13.0-71-generic
             system: Ubuntu 14.04 trusty
```

### PR DESCRIPTION
**Pull request by terminalmage**
## Summary

When a master-of-masters publishes a command, and a syndic master re-publishes it to its subscribed minions, it fires an event back to the master containing the list of minions matched by the syndic master for that job.

This pull request scans these events for lists of minions, and writes them to the job cache.

Fixes #30528.
## How to test
1. Use [this documentation](https://help.github.com/articles/checking-out-pull-requests-locally/) to check out this pull request into a new branch, and install from git.
2. Recommended setup: 3 hosts (1 master-of-masters, 1 syndic master, 1 lower-level minion). Set `order_masters: True` on the master-of-masters and subscribe the lower-level minion to the syndic master, and the syndic master to the master-of-masters.
3. From the master-of-masters, run `salt -v '*' test.ping`
4. Take the jid obtained in step 3 and run `salt-run jobs.list_job <jid>`. The `Minions` list in the return data should contain the minion ID of the lower-level minion. Prior to this pull request, the lower-level minions would not be in this return data. After completing the rest of the test steps, installing Salt from the head of `2015.8` on the master-of-masters and repeating steps 3 and 4 should show no lower-level minions in the `jobs.list_job` return data.
5. Stop the `salt-minion` service on the lower-level minion, and then repeat step 3.
6. Take the jid obtained in step 5 and run `salt-run jobs.lookup_jid <jid> missing=True`. The lower-level minion should be included in the return data, and should show `Minion did not return`. Repeating the call to `jobs.lookup_jid` without the `missing=True` should exclude the lower-level minion from the return data.
## Review notes

Of particular focus in the review of this pull request should be the following:
- [ ] The `couchbase_return` returner. To maintain API compatibility I needed to move the writing of minion lists into a separate function, as I did for the `local_cache` returner, and I did not have an instance to test this.
- [ ] The file locking code added in salt/utils/files.py.

### CODE REVIEW

**Comment by jtand** [comment]
Go Go Jenkins!

**Comment by jtand** [comment]
Go Go Jenkins!

**Comment by jtand** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
Go Go Jenkins!

File: salt/returners/etcd_return.py [hunk 1]
```diff
@@ -113,6 +113,13 @@ def save_load(jid, load):
     )
 
 
+def save_minions(jid, minions):  # pylint: disable=unused-argument
+    '''
+    Included for API consistency
+    '''
+    pass
```
Comment (cachedout) [inline salt/returners/etcd_return.py]: I'm sort of inclined to raise exceptions for these and catch and log higher up.

**Comment by cachedout** [comment]
(With apologies for being the person who brought up locking the first place.)

After looking at this, I'm wondering if instead of locking we consider splitting this write up into multiple files and then scanning those when we retrieve the minions list. (I'm talking about the local cache of course.) The locking approach is starting to worry me..

File: salt/returners/etcd_return.py (hunk 1 above)
Comment (terminalmage) [inline salt/returners/etcd_return.py]: The only reason that I did not is that we're not already saving the minion list for this returner.

**Comment by terminalmage** [comment]
@cachedout Good idea, I just pushed some changes, let me know what you think.

File: salt/returners/etcd_return.py (hunk 1 above)
Comment (cachedout) [inline salt/returners/etcd_return.py]: Fair enough. Good point.

**Comment by cachedout** [comment]
Yeah, I think this is better. Let's go with this.

**Pull request merged**